tests/unit/lib/secrets/vault_csi_tests.py

```diff
@@ -0,0 +1,205 @@
+    shutil.rmtree(old_data_path)
+
+
+def get_secrets_store(csi_dir: str) -> SecretsStore:
+    store = secrets_store_from_config({"secrets.path": csi_dir, "secrets.provider": "vault_csi"})
+    assert isinstance(store, VaultCSISecretsStore)
+    return store
+
+
+EXAMPLE_SECRETS_DATA = {
+    "secret/example-service/example-secret": {
+        "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+        "lease_id": "",
+        "lease_duration": 2764800,
+        "renewable": False,
+        "data": {"password": "password", "type": "credential", "username": "reddit"},
+        "warnings": None,
+    },
+    "secret/example-service/nested/example-nested-secret": {
+        "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+        "lease_id": "",
+        "lease_duration": 2764800,
+        "renewable": False,
+        "data": {"password": "password", "type": "credential", "username": "reddit"},
+        "warnings": None,
+    },
+    "secret/bare-secret": {
+        "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+        "lease_id": "",
+        "lease_duration": 2764800,
+        "renewable": False,
+        "data": {"password": "password", "type": "credential", "username": "reddit"},
+        "warnings": None,
+    },
+}
+
+EXAMPLE_UPDATED_SECRETS = EXAMPLE_SECRETS_DATA.copy()
+EXAMPLE_UPDATED_SECRETS.update(
+    {
+        "secret/example-service/example-secret": {
+            "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+            "lease_id": "",
+            "lease_duration": 2764800,
+            "renewable": False,
+            "data": {
+                "password": "new_password",
+                "type": "credential",
+                "username": "new_reddit",
+            },
+            "warnings": None,
+        },
+    }
+)
+
+
+class StoreTests(unittest.TestCase):
+    def setUp(self):
+        self.csi_dir = new_fake_csi(EXAMPLE_SECRETS_DATA)
+
+    def tearDown(self):
+        shutil.rmtree(self.csi_dir)
+
+    def test_can_load_credential_secret(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_credentials("secret/example-service/example-secret")
+        assert data.username == "reddit"
+        assert data.password == "password"
+
+    def test_symlink_updated(self):
+        original_data_path = self.csi_dir.joinpath("..data").resolve()
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_credentials("secret/example-service/example-secret")
+        gevent.sleep(0.1)  # prevent gevent from making execution out-of-order
+        assert data.username == "reddit"
+        assert data.password == "password"
+        simulate_secret_update(self.csi_dir)
+        assert original_data_path != self.csi_dir.joinpath("..data").resolve()
+        data = secrets_store.get_credentials("secret/example-service/example-secret")
+        assert data.username == "reddit"
+        assert data.password == "password"
```

Comment:
It's important to validate _more than one_ secret update -- historically, the main failure mode of our naive implementations that watch the file would work correctly for one update but will fail to notice the second.

Reply:
Done, also with secret value updates.